File: rules-runner/javascript/rules-runner-js/src/runners.ts
import { Rule, RuleEvaluationDataContext, RuleEvaluationResult, RuleSet, RuleSetEvaluationResult } from "./typings"
import { evaluate } from "certlogic-js"


export const runRule = (rule: Rule, data: RuleEvaluationDataContext): RuleEvaluationResult => {
    try {
        const evalResult = evaluate(rule.logic, data)
        if (typeof evalResult === "boolean") {
            return evalResult
        }
        return {
            errorMessage: `rule evaluated to a non-boolean: ${JSON.stringify(evalResult)}`
        }
    } catch (e) {
        return {
            errorMessage: `rule evaluation errored out: ${e.message}`
        }
    }
}

export const runRuleSet = (ruleSet: RuleSet, data: RuleEvaluationDataContext): RuleSetEvaluationResult => {
    const ruleEvaluations: { [ruleId: string]: RuleEvaluationResult } = {}
    ruleSet.forEach((rule) => {
        ruleEvaluations[rule.id] = runRule(rule, data)
    })
    const hasErrors = Object.values(ruleEvaluations).some((ruleResult) => typeof ruleResult !== "boolean")
    const allSatisfied = !hasErrors && Object.values(ruleEvaluations).map((ruleResult) => ruleResult as boolean).reduce((l, r) => l && r)
    return {
        ruleEvaluations,
        hasErrors,
        allSatisfied
    }
}

File: rules-runner/rules-runner-kotlin/src/main/kotlin/eu/ehn/dcc/rulesets/runners.kt
package eu.ehn.dcc.rulesets

import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.databind.node.BooleanNode
import eu.ehn.dcc.certlogic.evaluate
import eu.ehn.dcc.rulesets.RuleEvaluationResult.BooleanResult
import eu.ehn.dcc.rulesets.RuleEvaluationResult.EvaluationError


fun runRule(rule: Rule, data: JsonNode): RuleEvaluationResult =
    try {
        val evalResult = evaluate(rule.logic, data)
        if (evalResult is BooleanNode) {
            BooleanResult(evalResult.booleanValue())
        } else {
            EvaluationError("rule evaluated to a non-boolean: $evalResult")
        }
    } catch (re: RuntimeException) {
        EvaluationError("rule evaluation errored out: ${re.message}")
    }


fun runRuleSet(ruleSet: RuleSet, data: JsonNode): RuleSetEvaluationResult {
    val ruleEvaluations = mutableMapOf<String, RuleEvaluationResult>()
    ruleSet.forEach { rule ->
        ruleEvaluations.put(rule.id, runRule(rule, data))
    }
    val hasErrors = ruleEvaluations.values.any { result -> result is EvaluationError }
    val allSatisfied = !hasErrors && ruleEvaluations.values
        .map { ruleResult -> (ruleResult as BooleanResult).value }
        .fold(true) { acc, b -> acc && b }
    return RuleSetEvaluationResult(ruleEvaluations, hasErrors, allSatisfied)
}

